examples/autopilot_server/autopilot_server.cpp

```diff
@@ -0,0 +1,312 @@
+#include <mavsdk/plugins/param_server/param_server.h>
+#include <mavsdk/plugins/param/param.h>
+#include <mavsdk/plugins/telemetry_server/telemetry_server.h>
+#include <mavsdk/plugins/action_server/action_server.h>
+#include <mavsdk/plugins/mission_raw_server/mission_raw_server.h>
+#include <mavsdk/plugins/mission/mission.h>
+
+/*
+ This example runs a MAVLink "autopilot" utilising the MAVSDK server plugins
+ on a seperate thread. This uses two MAVSDK instances, one GCS, one autopilot.
+
+ The main thread acts as a GCS and reads telemetry, parameters, transmits across
+ a mission, clears the mission, arms the vehicle and then triggers a vehicle takeoff.
+
+ The autopilot thread handles all the servers and triggers callbacks, publishes telemetry,
+ handles and stores parameters, prints received missions and sets the vehicle height to 10m on
+ successful takeoff request.
+*/
+
+using namespace mavsdk;
+
+using std::chrono::seconds;
+using std::chrono::milliseconds;
+using std::this_thread::sleep_for;
+
+#define ERROR_CONSOLE_TEXT "\033[31m" // Turn text on console red
+#define TELEMETRY_CONSOLE_TEXT "\033[34m" // Turn text on console blue
+#define NORMAL_CONSOLE_TEXT "\033[0m" // Restore normal console colour
+
+Mission::MissionItem make_mission_item(
+    double latitude_deg,
+    double longitude_deg,
+    float relative_altitude_m,
+    float speed_m_s,
+    bool is_fly_through,
+    float gimbal_pitch_deg,
+    float gimbal_yaw_deg,
+    Mission::MissionItem::CameraAction camera_action)
+{
+    Mission::MissionItem new_item{};
+    new_item.latitude_deg = latitude_deg;
+    new_item.longitude_deg = longitude_deg;
+    new_item.relative_altitude_m = relative_altitude_m;
+    new_item.speed_m_s = speed_m_s;
+    new_item.is_fly_through = is_fly_through;
+    new_item.gimbal_pitch_deg = gimbal_pitch_deg;
+    new_item.gimbal_yaw_deg = gimbal_yaw_deg;
+    new_item.camera_action = camera_action;
+    return new_item;
+}
+
+int main(int argc, char** argv)
+{
+    // We run the server plugins on a seperate thead so we can use the main
+    // thread as a ground station.
+    std::thread autopilotThread([]() {
+        mavsdk::Mavsdk mavsdkTester;
+        mavsdk::Mavsdk::Configuration configuration(
+            mavsdk::Mavsdk::Configuration::UsageType::Autopilot);
+        mavsdkTester.set_configuration(configuration);
+
+        auto result = mavsdkTester.add_any_connection("udp://127.0.0.1:14551");
+        if (result == mavsdk::ConnectionResult::Success) {
+            std::cout << "Connected autopilot server side!" << std::endl;
+        }
+
+        auto prom = std::promise<std::shared_ptr<mavsdk::System>>{};
+        auto fut = prom.get_future();
+        mavsdkTester.subscribe_on_new_system([&mavsdkTester, &prom]() {
+            std::cout << "Discovered MAVSDK GCS" << std::endl;
+            auto system = mavsdkTester.systems().back();
+            prom.set_value(system);
+            mavsdkTester.subscribe_on_new_system(nullptr);
```

Comment:
You should unsubscribe first to prevent a race.

```suggestion
            mavsdkTester.subscribe_on_new_system(nullptr);
            prom.set_value(system);
```